NBitcoin/Block.cs

```diff
@@ -9,9 +9,39 @@
 using System.Collections.Generic;
 using System.IO;
 using System.Linq;
+using System.Security.Cryptography.X509Certificates;
 
 namespace NBitcoin
 {
+	public class BlockHeaderFactory: IBlockHeaderFactory
+	{
+		public IBlockHeader<ICoinStream> CreateNewBlockHeader()
+		{
+			return new BlockHeader();
+		}
+
+		public IBlockHeader<ICoinStream> CreateNewBlockHeader(byte[] bytes)
+		{
+			return new BlockHeader(bytes);
+		}
+
+		public IBlockHeader<ICoinStream> CreateNewBlockHeader(string hex)
+		{
+			return new BlockHeader(hex);
+		}
+	}
+
+	public interface IBlockHeaderFactory
+	{
+		IBlockHeader<ICoinStream> CreateNewBlockHeader();
+	}
+
+	public interface IBlockHeader<in TCoinStream> : ICoinSerializable<TCoinStream> where TCoinStream: ICoinStream

```

Comment:
I am not sure I understand whats `ICoinStream` for.

The notion of `IBlockHeader` is good I would be happier if it was an `abstract` called `BlockHeaderBase` but I guess it means many changes.

Reply:
I went the interface route instead of a base class to leave Bitcoin as the first class integrated coin in the lib. It also minimizes code changes like you said. ;)

While abstracting `IBitcoinSerializable` to allow different  `ReadWrite`  methods on a blockheader level I realized that it relied completely on `BitcoinStream` which may or may not be too coupled to btc to make all blockheaders utilize `BitcoinStream` in `ReadWrite`. I'm probably wrong though,  not very sure what that part does.

Reply:
I would agree if we could break existing code, but we want to prevent this, I think we should not try to abstract this and just subclass `BlockHeader` for the time being.

Reply:
I'll change this now. 

Maybe it's because I haven't used NBitcoin as extensively as you but I can't see why abstracting NBitcoin's BlockHeader would be breaking: Everyone can still use BlockHeader as before and it will still be the default, it's only when they try to use the new functionality that usage of the interface would be needed and visible.